### ISSUE TITLE
WriteFile not saving properly a file?

### ISSUE THREAD

**Issue by TimeMaster18** [rickyah/ini-parser#189]
Not sure if it's a bug or something I might be doing wrong, but using WriteFile seems to create some problems when reading the file later.

Seems to work fine if after the WriteFile from IniParser, you use File.WriteAllText.

**Comment by Xinnony** [comment]
Same problem, how solve this ?

**Comment by rickyah** [comment]
Hi, could you please be more specific on the  problems that you are finding?

**Comment by TimeMaster18** [comment]
Basically I followed the example over the wiki:

```
var parser = new FileIniDataParser();

//Get the data
IniData data = new IniData();

//Add a new section and some keys
data.Sections.AddSection("newSection");
data["newSection"].AddKey("newKey1", "value1");
data["newSection"].AddKey("newKey2", "value5");

//Save the file
parser.WriteFile("newINIfile.ini", data);
```

Result: the file is created but the program that reads it doesn't work, it crashes because of it. If I create it manually it works. If after that I use the following then it works:

`File.WriteAllText("newINIfile.ini", File.ReadAllText("newINIfile.ini"));`

Might be some formatting of the new lines or end of file, no idea.

**Comment by rickyah** [comment]
I can't reproduce the bug can you provide more insight on the issue?
- Version of IniParser you are using
- OS and version
- .Net Framework used (netcore, mono, .net framework) and version

For this specific case, a diff of both files (the one written by my example program and the file you write with `File.WriteAllText`) would be handy. You can just attach both files so I can do a diff if you please

This a weird bug because the writing operations are .NET function calls, so if there is a should be a very noticeable bug. There is, however one difference between the `parser.WriteFile` and your `File.WriteAllText`, according to the Remarks in the MSDN doc: both methods write the text encoded in UTF8, but mine inserts a BOM byte, but yours does not because you don't specify the encoding:

https://docs.microsoft.com/en-us/dotnet/api/system.io.file.writealltext?view=netframework-4.8#System_IO_File_WriteAllText_System_String_System_String_
> Remarks

> This method uses UTF-8 encoding without a Byte-Order Mark (BOM), so using the GetPreamble method will return an empty byte array. If it is necessary to include a UTF-8 identifier, such as a byte order mark, at the beginning of a file, use the WriteAllText(String, String, Encoding) method overload with UTF8 encoding.

**Comment by TimeMaster18** [comment]
I compared them with a hex editor, the only difference is the start of the file. With yours it has 3 bytes more at the very start, the rest is identical.

These are the 3 extra bytes at the start:

EF BB BF

The .ini is used by another application that isn't mine, so I can't get much more from it, error just says that the file is invalid.

EDIT: Those are indeed, the bytes you mention. Seems like the application using the ini doesn't like it.

**Comment by rickyah** [comment]
Yeah that’s the BOM. I’m afraid the fix for you is to save the IniData yourself using the file write methods that does NOT specify the encoding, as those methods still write the file in UTF8, but they do not write the BOM bytes.

**Comment by jbennink** [comment]
I know this is closed but I ran into this same issue where the ini file was UTF8 and after changing it with ini-parser it ended up UTF8-BOM. My external program did not like the extra bytes added. I followed the "solution above" with the extra read/write which solved it.
I also tried giving  it a fixed Encoding.UTF8 thrid paramater but that still saved it as UTF8 BOM. I then tried Encoding.ASCII which saved it as UCS-2 LE BOM (according to Notepad++) and my program did accept this without problems.

1. Should Encoding.UTF8 not save it as plain UTF8 encoding?
2. Would it be possible to detect the encoding on loading the ini file and reusing that encoding for the save by default.
